### PR TITLE
added Harness CI/CD tutorial

### PR DESCRIPTION
**Pull request by pavanbelagatti**
# Harness Developer Pull Request
Thanks for helping us make the Developer Hub better. The PR will be looked at
by the maintainers. 

## What Type of PR is This?

- [ ] Issue
- [ ] Feature
- [ ] Maintenance/Chore

If tied to an Issue, list the Issue(s) here:

* *Issue(s)*

## House Keeping
Some items to keep track of. Screen shots of changes are optional but would help the maintainers review quicker. 

- [ ] Tested Locally
- [ ] *Optional* Screen Shoot.

### CODE REVIEW

File: tutorials/deploy-services/harness-cicd-tutorial.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
+# Harness Continuous Integration and Continuous Delivery
```
Comment (ravilach) [inline tutorials/deploy-services/harness-cicd-tutorial.md:1]: Can we change the title e.g the H1 Tag here to `# Build and Deploy a NodJS App to Kubernetes`

File: tutorials/deploy-services/harness-cicd-tutorial.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+Software development teams that adopt a CI/CD pipeline see faster software deployments with more negligible risk and quicker bug identification and remediation. If you are new to CI/CD, it can seem like an overwhelming amount of new terms and processes to learn. But setting up a CI/CD pipeline is easier than you think. 
+
+A CI/CD pipeline helps your team pinpoint where problems in your software are likely to occur – before they reach end users. Additionally, it allows you to release updates more frequently and spend less time testing individual components or modules before releasing them to users. Harness is a leader in CI/CD with impeccable features. Today, we will show you how to implement CI/CD in minutes with a simple setup. 
+![CI/CD flowchart](./static/harness-cicd-tutorial/harness_cicd_pipeline.png)
+
+### Pre-Requisites
+- Free [Harness](https://app.harness.io/auth/#/signup/?utm_source=internal&utm_medium=social&utm_campaign=community&utm_content=pavan_cicd_article&utm_term=get-started) account to do CI/CD (on-premise)
```
Comment (ravilach) [inline tutorials/deploy-services/harness-cicd-tutorial.md:46]: Can we change the UTM links to `https://app.harness.io/auth/#/signup/?module=cd&?utm_source=website&utm_medium=harness-developer-hub&utm_campaign=cd-plg&utm_content=get-started`

**Review by ravilach — CHANGES_REQUESTED** [review]

**Comment by ravilach** [comment]
@pavanbelagatti thanks for the contribution. You can just re-commit to your branch for the changes. Tagging in @michaelcretzman as the CD reviewer. Cheers!

File: tutorials/deploy-services/harness-cicd-tutorial.md (hunk 1 above)
Comment (pavanbelagatti) [reply tutorials/deploy-services/harness-cicd-tutorial.md]: Updated the Header Title

File: tutorials/deploy-services/harness-cicd-tutorial.md (hunk 2 above)
Comment (pavanbelagatti) [reply tutorials/deploy-services/harness-cicd-tutorial.md]: Changed the link! Thanks for the quick feedback @ravilach

**Comment by pavanbelagatti** [comment]
@michaelcretzman can you please take a look

File: tutorials/deploy-services/harness-cicd-tutorial.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+`docker buildx build --platform=linux/arm64 --platform=linux/amd64  -t docker.io/<docker hub username>/<image name>:<tag> --push  -f ./Dockerfile .`
+
+Once the build and push are successful, you can confirm it by going to your Docker Hub account.
+![DockerHub image](./static/harness-cicd-tutorial/hello_world_docker_image.jpeg)
+
+You can see the `deployment.yaml` file in the forked repo, which defines the deployment yaml file to help us deploy the application to our Kubernetes cluster. At this point, make sure your Kubernetes cluster is up and running.
+
+Once everything is set, it is time to set up a Harness account to do CI/CD. Create a free Harness account and your first project. Once you sign-up at Harness, you will be presented with the new CI/CD experience and capabilities.
+
+Add the required connectors, GitHub repo, Docker Hub and secrets, if any. Delegate in Harness is a service/software you need to install/run on the target cluster [Kubernetes cluster in our case] to connect your artifacts, infrastructure, collaboration, verification and other providers with the Harness Manager. When you set up Harness for the first time, you install a Harness Delegate.
```
Comment (KimberlyFields) [inline tutorials/deploy-services/harness-cicd-tutorial.md:64]:
```suggestion
4. Add the required connectors, GitHub repo, Docker Hub and secrets, if any. The Harness Delegate is a lightweight worker process that you need to install on the target cluster to connect your artifacts, infrastructure, collaboration, verification, and other providers with the Harness Manager. When you set up Harness for the first time, you install a Harness Delegate.
```

File: tutorials/deploy-services/harness-cicd-tutorial.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+![Harness Connectors](./static/harness-cicd-tutorial/project_connectors.jpeg)
+
+Select the Continuous Integration module and add the necessary stages and steps, as shown below.
```
Comment (KimberlyFields) [inline tutorials/deploy-services/harness-cicd-tutorial.md:67]:
```suggestion
5. Select the Continuous Integration module, and then add the necessary stages and steps, as shown below.
```

Comment (KimberlyFields) [inline tutorials/deploy-services/harness-cicd-tutorial.md:67]:
```suggestion
5. Select the Continuous Integration module, and then add the necessary stages and steps, as shown below.
```

File: tutorials/deploy-services/harness-cicd-tutorial.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+![cicd with harness images](./static/harness-cicd-tutorial/build_and_test_stage.jpeg)
+
+‘Test’ step set-up is done as below,
```
Comment (KimberlyFields) [inline tutorials/deploy-services/harness-cicd-tutorial.md:70]:
```suggestion
6. Set up the Test step as shown below.
```

File: tutorials/deploy-services/harness-cicd-tutorial.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+![npm test](./static/harness-cicd-tutorial/configure_run_step.jpeg)
+
+The “‘Push to Docker Registry” step is as below,
```
Comment (KimberlyFields) [inline tutorials/deploy-services/harness-cicd-tutorial.md:73]:
```suggestion
7. Set up the Push to Docker Registry step as shown below.
```

File: tutorials/deploy-services/harness-cicd-tutorial.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+![build and push image](./static/harness-cicd-tutorial/build_push_image.png)
+
+Next, set up a deployment pipeline.
```
Comment (KimberlyFields) [inline tutorials/deploy-services/harness-cicd-tutorial.md:76]:
```suggestion
8. Next, set up a deployment pipeline.
```

File: tutorials/deploy-services/harness-cicd-tutorial.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+![CD overview](./static/harness-cicd-tutorial/deployment_pipeline.jpeg)
+
+Add the required details in the “Service” tab.
```
Comment (KimberlyFields) [inline tutorials/deploy-services/harness-cicd-tutorial.md:79]:
```suggestion
9. Add the required details on the **Service** tab.
```

File: tutorials/deploy-services/harness-cicd-tutorial.md [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+![deploy service](./static/harness-cicd-tutorial/service_tab.png)
+
+Define the environment type in the “Environment” tab.
```
Comment (KimberlyFields) [inline tutorials/deploy-services/harness-cicd-tutorial.md:82]:
```suggestion
10. Define the environment type in the **Environment** tab.
```

File: tutorials/deploy-services/harness-cicd-tutorial.md [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+![environment](./static/harness-cicd-tutorial/environment_tab.png)
+
+Strategise the execution by selecting which deployment you prefer.
```
Comment (KimberlyFields) [inline tutorials/deploy-services/harness-cicd-tutorial.md:85]:
```suggestion
11. Strategize the execution by selecting the deployment you prefer.
```

File: tutorials/deploy-services/harness-cicd-tutorial.md [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+![execution](./static/harness-cicd-tutorial/execution_tab.jpeg)
+
+Save everything and run the pipeline.
```
Comment (KimberlyFields) [inline tutorials/deploy-services/harness-cicd-tutorial.md:88]:
```suggestion
12. Save everything, and then run the pipeline.
```

File: tutorials/deploy-services/harness-cicd-tutorial.md [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+![run pipeline](./static/harness-cicd-tutorial/run_pipeline.jpeg)
+
+You can see both CI and CD getting executed one by one with all the steps specified.
```
Comment (KimberlyFields) [inline tutorials/deploy-services/harness-cicd-tutorial.md:91]:
```suggestion
You can see both CI and CD being executed one by one with all the steps specified.
```

File: tutorials/deploy-services/harness-cicd-tutorial.md [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+![CI/CD execution](./static/harness-cicd-tutorial/ci_cd_execution_successful.png)
+
+**Congratulations!** We successfully built, and tested the application code and deployed it onto our Kubernetes cluster using the Harness platform.
```
Comment (KimberlyFields) [inline tutorials/deploy-services/harness-cicd-tutorial.md:94]:
```suggestion
**Congratulations!** You successfully built and tested the application code and deployed it onto the Kubernetes cluster.
```

File: tutorials/deploy-services/harness-cicd-tutorial.md [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+
+You can confirm this deployment by using the command `kubectl get pods`
+![kubernetes deployments](./static/harness-cicd-tutorial/pods_running.jpeg)
+
+You can see two replicas running as per our specifications on the `deployment.yaml` file. Also, confirm the same by going to your Kubernetes dashboard. Since I am using Google cloud (GCP), I can see and confirm that there are two pods running.
```
Comment (KimberlyFields) [inline tutorials/deploy-services/harness-cicd-tutorial.md:99]:
```suggestion
You can see two replicas running as per the specifications on the `deployment.yaml` file. Also, confirm the same by going to your Kubernetes dashboard. The following image shows two pods running on Google cloud (GCP).
```

**Review by KimberlyFields — CHANGES_REQUESTED** [review]
Be sure to use the **Add suggestion to batch** option to accept the suggestions instead of committing each suggestion individually. The batch option combines all the changes into a single commit. The preview build is rebuilt every time there's a new commit, so doing a batch prevents the preview build from rebuilding multiple times.

**Review by KimberlyFields — APPROVED** [review]

**Comment by KimberlyFields** [comment]
@ravilach you requested some changes on this PR and it looks like they were addressed. Are you ok with this PR being merged?

**Comment by michaelcretzman** [comment]
PR gives me a 404. 
All links to the MD file here are outdated.
Removing myself as reviewer.

**Comment by ravilach** [comment]
Sorry I missed this convo. @KimberlyFields looks good to me. @pavanbelagatti we will need to get your tile added to https://developer.harness.io/tutorials/deploy-services, can create a separate PR for that.

**Pull request merged**